### ISSUE TITLE
Change image loading strategy

### ISSUE THREAD

**Issue by madoar** [PhoenicisOrg/phoenicis#1616]
Currently we load images inside Java code with the classloader belonging to the class which requires/uses the loaded images.
This approach, requires image paths relative to the package of the class using an image.
These image paths are likely to be quite deep.
One example for such an image path are the images located in `org/phoenicis/javafx/components/common/widgets/utils`, which are loaded by the `ListWidgetElement` class.

I believe we should think about changing our approach to load images from inside java code to allow flatter folder structures.
Any ideas?

**Comment by plata** [comment]
No idea. Especially with Java modules, you should use the class loader (there's no classpath anymore).

**Comment by madoar** [comment]
Yes, but the class loader typically uses the classpath underneath to locate the correct resource file.
What we could maybe use is the [Resources](https://google.github.io/guava/releases/21.0/api/docs/com/google/common/io/Resources.html) class from guava.

**Comment by plata** [comment]
Where would you like to place the images instead?

**Comment by madoar** [comment]
That is the question^^
I think we should separate the image folder structure from our package structure.
This means, we can remove the `org/phoenicis/javafx` part and we can also shorten the later paths.
For the given example above I think something like `widgets/selector` or `list-widgets/selector` would be good.

**Comment by madoar** [comment]
I don't think this is a problem.
The [source code](https://github.com/google/guava/blob/master/guava/src/com/google/common/io/Resources.java#L190) of `Resources` shows that they use a `ClassLoader` instance underneath, which is the same as we do.

**Comment by qparis** [comment]
So to continue the discussion:

If you remove the org/phoenicis/javafx parts, it means that the resources will be directly put on the root of the compiled jar file right? 
If it is the case, it means that the resources are no longer namespaced. If our dependency do the same, we could have conflicts

**Comment by qparis** [comment]
It is not just a convention. It allows to differentiate files from different namespaces (exactly what it does for compiled .class files).

In the following example, if you put javafx.properties directly in the resource folder and one of your dependencies does the same with the same name of file, you won't be able to predict what file you have at the end 

![image](https://user-images.githubusercontent.com/373645/51799984-04244480-2229-11e9-9d87-5890959561fb.png)

**Comment by madoar** [comment]
I fully agree with you that the naming scheme is used to prevent file name conflicts. Still it is mainly a convention in my eyes. In my opinion every convention has an aim, which specifies why you should use the convention, for this convention the aim is to prevent file name conflicts.

Still we need to decide whether the prevention of a possible filename conflict or a easier to use, understand and maintain folder structure is more important for us.

**Comment by qparis** [comment]
I am really not convinced that removing "org" and "phoenicis" directory will lead to a more maintenable folder structure. I think it can potentially create problems however

**Comment by qparis** [comment]
I mean, why using this namespace for java file in that case

**Comment by madoar** [comment]
`org/phoenicis` is not really problematic, the later path parts are in my opinion.
I think we should save all images in a folder, that belong together according to their functionality, and not according to the class in which they are used.

**Comment by qparis** [comment]
The classes should already been grouped by functionnality (otherwise, we should refactor the code so that it is more consistent). 

The corner case is when we have resources that are used by several classes. Then I agree that we could create a dedicated directory (Even though it probably mean that our code is not structured well)

**Comment by madoar** [comment]
> The classes should already be grouped by functionnality (otherwise, we should refactor the code so that it is more consistent).

Yes, but only on a package scope. There may exist subpackages which contain several specialization classes. These subpackages would also be reflected in the resources paths. This is not always a good idea.

**Comment by qparis** [comment]
Then, the resource should be placed on the most common parent package

**Comment by madoar** [comment]
Yes, but how do we access it then from inside the subpackage without using absolute locations?

**Comment by qparis** [comment]
Depend on the case, but we probably should not access generic resources in specific classes most of the cases

**Comment by madoar** [comment]
What do you mean by `generic resources`? I'm thinking about specialized resources like an icon for a type of specific type class which is located in a subpackage with its supporting classes.

**Comment by qparis** [comment]
You could for example imagine a pattern when you have an helper class that will provide you the icons

**Comment by madoar** [comment]
Isn't this basically what `Resources` from `guava` does? How would your helper class be used?

**Comment by qparis** [comment]
I really don't like the idea of adding a heavy dependency just for this particular case
